File: java/NeetCode150/14_rain_water.java
class Solution 
{
    public int trap(int[] height) 
    {
        if(height.length < 3) return 0; // Can't catch water without atleast 3 elements in the array
        int l = 0;
        int end = height.length - 1;
        
        // Get rid of the uneeded height
        while(height[l] == 0) {l += 1;}
        while(height[end] == 0) {end -= 1;}

        while(l > 0 && height[l+1] > height[l]) {l+=1;}
        while(height[end-1] > height[end]) {end-=1;}

        // Removed the height not needed, start true logic.
        int r = l + 1;
        int total_water = 0;
        while (r < end)
        {
            while(height[l] > height[r] && r < end){r+=1;}
            while (Math.max(height[l+1],  height[r]) > height[r]) {l+=1;}
            int min_height = Math.min(height[l], height[r]);
            for(int i = l+1; i<r;i++)
            {
                total_water += min_height - height[i];
            }
            l = r;
            r += 1;
        }
        return total_water;
    }
}


// l = 1
// r = 2
// end = 7

File: python/NeetCode150/14_rain_water.py
class Solution:
    def trap(self, heights: List[int]) -> int:
        l = 0
        end = len(heights) - 1
        total_water = 0
        if len(heights) < 2:
            return 0

        # Remove any leading or trailing 0s
        while (heights[l] == 0):
             l+=1
        while (heights[end] == 0):
             end-=1

        # Remove any stairs that water could not hold.
        while (heights[l+1] >= heights[l]):
             l+=1
        while (heights[end-1] >= heights[end]):
            end-=1
        
        # To the right of l and the the left of r there will have to be water.
        r = l + 1
        while (r < end):
            while heights[l] > heights[r] and r < end: #
                r += 1
            while max(heights[l+1],  heights[r]) > heights[r]:
                l += 1
            min_height = min(heights[l], heights[r]) # Water lvl
            for wata in range(l+1, r):
                total_water += min_height - heights[wata]
            l = r
            r+=1
        return total_water
            
             
            


